test/unit/puppetlabs/jruby_utils/lockable_ref_pool_test.clj

```diff
@@ -0,0 +1,638 @@
+      (.lock pool)
+      (is (.isLocked pool))
+      (let [borrow-thread-started-1? (promise)
+            borrow-thread-started-2? (promise)
+            borrow-thread-borrowed-1? (promise)
+            borrow-thread-borrowed-2? (promise)
+            borrow-thread-1 (future (deliver borrow-thread-started-1? true)
+                                    (let [instance (.borrowItem pool)]
+                                      (deliver borrow-thread-borrowed-1? true)
+                                      (.releaseItem pool instance))
+                                    true)
+            borrow-thread-2 (future (deliver borrow-thread-started-2? true)
+                                    (let [instance (.borrowItem pool)]
+                                      (deliver borrow-thread-borrowed-2? true)
+                                      (.releaseItem pool instance))
+                                    true)]
+        @borrow-thread-started-1?
+        @borrow-thread-started-2?
+        (is (not (realized? borrow-thread-borrowed-1?)))
+        (is (not (realized? borrow-thread-borrowed-2?)))
+
+        (let [instance (.borrowItem pool)]
+          (is (true? true))
+          (.releaseItem pool instance))
+
+        (is (true? true))
+        (.unlock pool)
+        (is (true? (timed-deref borrow-thread-1))
+            "timed out waiting for first borrow thread to finish")
+        (is (true? (timed-deref borrow-thread-2))
+            "timed out waiting for second borrow thread to finish"))
+        (is (not (.isLocked pool))))))
+
+(deftest pool-lock-reentrant-for-many-locks-test
+  (testing "multiple threads cannot lock the pool while it is already locked"
+    (let [pool (create-populated-pool 1)]
+      (is (not (.isLocked pool)))
+      (.lock pool)
+      (is (.isLocked pool))
+      (let [lock-thread-started-1? (promise)
+            lock-thread-started-2? (promise)
+            lock-thread-locked-1? (promise)
+            lock-thread-locked-2? (promise)
+            lock-thread-1 (future (deliver lock-thread-started-1? true)
+                                  (.lock pool)
+                                  (deliver lock-thread-locked-1? true)
+                                  (.unlock pool)
+                                  true)
+            lock-thread-2 (future (deliver lock-thread-started-2? true)
+                                  (.lock pool)
+                                  (deliver lock-thread-locked-2? true)
+                                  (.unlock pool)
+                                  true)]
+        @lock-thread-started-1?
+        @lock-thread-started-2?
+        (is (not (realized? lock-thread-locked-1?)))
+        (is (not (realized? lock-thread-locked-2?)))
+        (.unlock pool)
+        (is (true? (timed-deref lock-thread-1))
+            "timed out waiting for first lock thread to finish")
+        (is (true? (timed-deref lock-thread-2))
+            "timed out waiting for second lock thread to finish"))
```

Comment:
Will these two threads race to acquire the lock once the main thread unlocks? Maybe that doesn't even matter? I guess I'm wondering why two lock threads are even involved in this test?

Reply:
I'm not sure why there are multiple, though the test description makes it seem like that was very intentional. They might race in that it's undetermined who will acquire the lock first, but since both threads also immediately unlock, whoever gets it first will give it up right away, allowing the other one to proceed. So by the time the second test gets evaluated, both are guaranteed to be done (either the first will complete immediately, or it will block till Thread 2 gives up the lock and Thread 1 acquires it, and then succeed that way).

Reply:
yeah, I think that's the conclusion I came to, but was still trying to understand the value of testing with two threads.